### ISSUE TITLE
Unecessary `mut` reference in parameter

### ISSUE THREAD

**Issue by emilk** [rust-lang/rust-clippy#8863]
### What it does

It warns against using `&mut` in an parameter where `&` would suffice.

### Lint Name

`unnecessary-mut-ref`

### Category

suspicious

### Advantage

Less `mut` means less fights with the borrow checker. It can also lead to more opportunities for parallelization.

This lint could also catch bugs where the user intended to mutate an argument but forgot too.

### Drawbacks

A user may want to create an API where they reserve the right to mutate an argument, but does not do so yet.

We should not warn when implementing a `trait`.

### Example

```rust
fn foo(&mut self, y: &mut i32) -> i32 {
    self.x + *y
}
```

Could be written as:

```rust
fn foo(&self, y: &i32) -> i32 {
    self.x + *y
}
```

**Comment by junderw** [comment]
I wonder if this could be also made into a separate lint where ownership is passed when a `&mut` or `&` would suffice.

Basically, warning when your fn is taking more than it needs.

I like it!

**Comment by Centri3** [comment]
> I wonder if this could be also made into a separate lint where ownership is passed when a `&mut` or `&` would suffice.
> 
> Basically, warning when your fn is taking more than it needs.
> 
> I like it!

Is this already covered by `needless_pass_by_value` ?

**Comment by junderw** [comment]
Good point.

Maybe instead of `unnecessary-mut-ref` it should be proposed that this issue's lint is `needless_pass_by_mut_ref`

**Comment by junderw** [comment]
Referencing the pass_by_value PR in case someone wants to look at it for inspiration: #1556

**Comment by xTibor** [comment]
This lint could also suggest renaming the function if it has a `_mut` suffix, just to ensure the suggested code conforms the Rust naming conventions:

```rust
fn foo_mut(&mut self, y: &mut i32) -> i32;
// =>
fn foo(&self, y: &i32) -> i32;
```

**Comment by GuillaumeGomez** [comment]
I'll take a look.

**Comment by GuillaumeGomez** [comment]
I'm done with the implementation. I'll open a draft PR after the cleanup but I'll need people to test it so I can add more cases into the UI test.

**Comment by GuillaumeGomez** [comment]
I opened the draft [here](https://github.com/rust-lang/rust-clippy/pull/10900). Please test it so we can uncover cases I didn't think about. Thanks in advance! :)
